### PR TITLE
Support fuzzy-matching record snapshots

### LINKED ISSUE #9430: Enhance the `HapiSpec` framework with a record-snapshot facility
Introduce a `HapiSpecOperation` named `withSnapshot(SnapshotMode)` where
```
enum SnapshotMode { TAKE_FROM_STANDALONE_STREAMS, ASSERT_FOR_HAPI_TEST }
```

Support both (1) persisting a snapshot in a JSON artifact in the repository when running against mono-service; and (2) validating a repeat run as `@HapiTest` against that snapshot via fuzzy-matching.

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #9430
 - Adds a `SnapshotModeOp` that either,
    1. Saves a JSON snapshot of the records generated by a `HapiSpec` into a _record-snapshots/_ directory; or,
    2. Fuzzy-matches the records generated against an existing snapshot.
 - For example, the snapshot [here](https://github.com/hashgraph/hedera-services/blob/f449825665373e0a09465c29ab11fc3e5ed3b33c/hedera-node/test-clients/record-snapshots/CryptoTransfer-okToRepeatSerialNumbersInBurnList.json) was taken by running `CryptoTransferSuite.okToRepeatSerialNumbersInBurnList()` against mono-service with `.given(snapshotMode(TAKE_FROM_MONO_STREAMS), ...)`.
 - And the checked-in code [here](https://github.com/hashgraph/hedera-services/blob/09430-hapispec-snapshots/hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/CryptoTransferSuite.java#L288) with `.given(snapshotMode(FUZZY_MATCH_AGAINST_HAPI_TEST_STREAMS), ...)` means that whenever this spec runs as a `@HapiTest`, its generated records will be fuzzy-matched against the mono-service snapshot.
 - So before @mhess-swl merged [this PR](https://github.com/hashgraph/hedera-services/commit/3c3c8b2b9ea8bea9c77ca5b016830dbdd2eb0bb6) this was failing with a message like,
```
    2023-10-27 14:55:19.156 ERROR  526  HapiSpec - Record snapshot fuzzy-match failed
        org.opentest4j.AssertionFailedError: Mismatched field counts between expected senderAccountID {
          accountNum: 0
        }
        receiverAccountID {
          accountNum: 1005
        }
        serialNumber: 1
         and serialNumber: 1
  ...
```

...because until that PR, the `senderAccountID` and `receiverAccountID` were not being set in the `TokenTransferList` on a mint.

### CODE REVIEW

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM ! Thanks @tinker-michaelj .

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+                Assertions.fail(
+                        "Mismatched field names ('" + expectedName + "' vs '" + actualName + "' between expected "
+                                + expectedMessage + " and " + actualMessage + " - " + mismatchContext.get());
+            }
+            if (FIELDS_TO_SKIP_IN_FUZZY_MATCH.contains(expectedName)) {
+                continue;
+            }
+            final var expectedValue = expectedField.getValue();
+            final var actualValue = actualField.getValue();
+            if (expectedValue instanceof List<?> expectedList) {
```
Comment (david-bakin-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java:355]: these nested `if`s look so much like `matchValue` I'm wondering if L355-L384 should be broken out to a separate method to make that more obvious?  

probably doesn't matter

**Review by david-bakin-sl — DISMISSED** [review]
Hard to believe we've come this far with our HAPI tests _without_ this facility!

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+     */
+    private long placeholderAccountNum;
+    /**
+     * The location of the record stream to snapshot or fuzzy-match against.
+     */
+    private String recordsLoc;
+    /**
+     * The location of the record stream to snapshot or fuzzy-match against.
+     */
+    private String snapshotLoc;
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java]: Nope, fixed!

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+        return false;
+    }
+
+    /**
+     * Returns whether this operation has work to do, i.e., whether it could run against the target network.
+     *
+     * @return if this operation can run against the target network
+     */
+    public boolean hasWorkToDo() {
+        return fullSpecName != null;
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java:219]: Added comment clarifying that we leave this `null` if the target network doesn't match the snapshot mode.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/domain/EncodedItem.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import com.hederahashgraph.api.proto.java.TransactionRecord;
+import java.util.Base64;
+
+/**
+ * A POJO for Jackson to use in storing a Base64-encoded {@code (TransactionBody, TransactionRecord)} pair.
+ */
+public class EncodedItem {
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/domain/EncodedItem.java:29]: Added

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordAssertions.java [hunk 5]
```diff
@@ -79,6 +81,16 @@ public static void triggerAndCloseAtLeastOneFile(final HapiSpec spec) throws Int
         allRunFor(spec, triggerOp);
     }
 
+    public static void uninterruptiblyTriggerAndCloseAtLeastOneFile(final HapiSpec spec) {
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/RecordAssertions.java]: Haha good point, renamed to `triggerAndCloseAtLeastOneFileIfNotInterrupted()`. 😁

File: hedera-node/test-clients/record-snapshots/CryptoTransfer-okToRepeatSerialNumbersInBurnList.json [hunk 6]
```diff
@@ -0,0 +1 @@
...jGgFkGgFlGgFmGgFn","b64Record":"CisIFiocCgwIARAMGgYIgK6ZpA8SDAgBEA8aBgiArpmkD1gHcgcBAgMEBQYHEjAXYxh/iV3idn9kYlAfhlx0D530HYsMjMJb06V5g1NyP4DnOx8LKZX0OUpZ6lHVZ78aDAjK8OmpBhDrgIetAyIPCgkIjvDpqQYQtQcSAhgCKiDDg8Kuw4LCt8ODwrl0RjjDgsKuSsODwovDg8KQw4PCjlIAWmAKAxjuBxoLCgIYABIDGO0HGAEaCwoCGAASAxjtBxgCGgsKAhgAEgMY7QcYAxoLCgIYABIDGO0HGAQaCwoCGAASAxjtBxgFGgsKAhgAEgMY7QcYBhoLCgIYABIDGO0HGAc="},{"b64Body":"Cg8KCQiP8OmpBhC5BxICGAISAhgDGNPwUyICCHgyIMODwq7DgsK3w4PCuXRGOMOCwq5Kw4PCi8ODwpDDg8KOsgIOCgMY7gcaBwEBAgMEBQY=","b64Record":"CiIIFiocCgwIARAMGgYIgK6ZpA8SDAgBEA8aBgiArpmkD1gBEjBaSnXWTbRRztCWdZXg3rUfFEuvtEXZmDsxgxnbPZ2XpkPBeMpx+2GH+U2ng3xC7AsaDAjL8OmpBhCLoOm8ASIPCgkIj/DpqQYQuQcSAhgCKiDDg8Kuw4LCt8ODwrl0RjjDgsKuSsODwovDg8KQw4PCjlIAWlMKAxjuBxoLCgMY7QcSAhgAGAEaCwoDGO0HEgIYABgCGgsKAxjtBxICGAAYAxoLCgMY7QcSAhgAGAQaCwoDGO0HEgIYABgFGgsKAxjtBxICGAAYBg=="},{"b64Body":"Cg8KCQiP8OmpBhC7BxICGAISAhgDGMKgUyICCHgyIMODwq7DgsK3w4PCuXRGOMOCwq5Kw4PCi8ODwpDDg8KOsgIICgMY7gcaAQc=","b64Record":"CiAIFiocCgwIARAMGgYIgK6ZpA8SDAgBEA8aBgiArpmkDxIwdSpiOk1hJ0EoGkmS0fQVweTC5f5kI7u3b++yO14AcgI1XqIqEREvBR3vrow8ZaPhGgwIy/DpqQYQ64GJyAMiDwoJCI/w6akGELsHEgIYAiogw4PCrsOCwrfDg8K5dEY4w4LCrkrDg8KLw4PCkMODwo5SAFoSCgMY7gcaCwoDGO0HEgIYABgH"}]}
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/record-snapshots/CryptoTransfer-okToRepeatSerialNumbersInBurnList.json:1]: Added a [README](https://github.com/hashgraph/hedera-services/blob/09430-hapispec-snapshots/hedera-node/test-clients/record-snapshots/README.md).

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+        }
+    }
+
+    private void writeSnapshotOf(@NonNull final List<ParsedItem> postPlaceholderItems) throws IOException {
+        final var recordSnapshot = new RecordSnapshot();
+        recordSnapshot.setPlaceholderNum(placeholderAccountNum);
+        final var encodedItems = postPlaceholderItems.stream()
+                .map(item -> EncodedItem.fromParsed(item.itemBody(), item.itemRecord()))
+                .toList();
+        recordSnapshot.setEncodedItems(encodedItems);
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java]: Made a `RecordSnapshot.from()` factory.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+            @NonNull final Supplier<String> mismatchContext) {
+        final var expectedType = expectedMessage.getClass();
+        final var actualType = actualMessage.getClass();
+        if (!expectedType.equals(actualType)) {
+            Assertions.fail("Mismatched types between expected " + expectedType + " and " + actualType + " - "
+                    + mismatchContext.get());
+        }
+        expectedMessage = normalized(expectedMessage, expectedPlaceholderNum);
+        actualMessage = normalized(actualMessage, actualPlaceholderNum);
+        // getAllFields() returns a TreeMap so ordering is deterministic here
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java]: Weakened comment to say we're guaranteed just a `SortedMap`.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/SnapshotModeOp.java]: Renamed `matchValue()` to `matchSingleValue()` and extracted this as `matchValue()`.

**Review by david-bakin-sl — APPROVED** [review]
LGTM

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! 🙌

**Pull request merged**